### PR TITLE
Fix py2->3 string issues

### PR DESCRIPTION
**Pull request by foozlevazquez**
* Handle exceptions with binary string content. Added test.
* Handle api_requests with binary content correctly in Python 3. Added test.

### CODE REVIEW

**Comment by foozlevazquez** [comment]
See #671 for the full gory history.

**Comment by foozlevazquez** [comment]
@dhermes Can you help me understand why [this](https://coveralls.io/builds/1969041/source?filename=gcloud%2Fstorage%2Fconnection.py#L239) is claiming no coverage, when it is being tested by [these tests](https://github.com/GoogleCloudPlatform/gcloud-python/blob/9066b7e75026cd55bb7be96afc988f5ec0b2a639/gcloud/storage/test_connection.py#L199-L217) with and without binary strings?   I'm hoping understanding that one will help me understand the other coverage failure.  Thanks.

**Comment by dhermes** [comment]
From https://travis-ci.org/GoogleCloudPlatform/gcloud-python/builds/51608518#L383

```
gcloud.datastore.transaction                 32      0      4      0   100%   
gcloud.exceptions                            78      1     14      1    98%   175
gcloud.storage                               38      0     12      0   100%   
gcloud.storage._helpers                      60      0     12      0   100%   
gcloud.storage._implicit_environ              5      0      0      0   100%   
gcloud.storage.acl                          152      0     38      0   100%   
gcloud.storage.blob                         153      0     28      0   100%   
gcloud.storage.bucket                       180      0     48      0   100%   
gcloud.storage.connection                    68      1     18      1    98%   239
```

It looks like you've got 2 branch misses.

----

The line numbers are given, so hopefully they are easy to find. Typically a branch miss like this means in code like

```python
if foo == bar:
    baz()
```

`foo` is always equal to `bar`. So you need to add a test for the "other" branch where `foo != bar`.

----

You can see this report by running `tox -e cover` locally. Sometimes a branch miss doesn't get the line reported, and in this case I generate the HTML report

```
coverage html
ls htmlcov/
```

File: gcloud/exceptions.py [hunk 1]
```diff
@@ -171,6 +171,9 @@ def make_exception(response, content, use_json=True):
     :rtype: instance of :class:`GCloudError`, or a concrete subclass.
     :returns: Exception specific to the error response.
     """
+    if not hasattr(content, 'keys') and not isinstance(content, str):
```
Comment (dhermes) [inline gcloud/exceptions.py]: This is not necessary. Just change

```python
if isinstance(content, str):
```

(my error on committing that) to

```python
if isinstance(content, six.string_types):
```

File: gcloud/storage/connection.py [hunk 2]
```diff
@@ -235,6 +235,9 @@ def api_request(self, method, path, query_params=None,
         if not 200 <= response.status < 300:
             raise make_exception(response, content)
 
+        if not isinstance(content, str):
```
Comment (dhermes) [inline gcloud/storage/connection.py]: This needs a comment.

I am very wary of using `str` to intentially mean `bytes` in Py2 and `unicode` in Py3. We should instead use things like `six.binary_type` and `six.text_type`.

**Comment by dhermes** [comment]
It's unclear which of these fixes were needed.

There are only two uses of `http.request` in our code:

```
$ git log -1 --pretty=%H  # Current HEAD
56f60bb94b313c397f5aeadee1f0e1f6e5b5290c
$ git grep -n 'http\.request' | egrep -v test | egrep -v -e '^_gcloud_vendor'
gcloud/datastore/connection.py:80:        headers, content = self.http.request(
gcloud/storage/connection.py:169:        return self.http.request(uri=url, method=method, headers=headers,
```

The auth issue should be fixed upstream in `oauth2client`.

The only change here is one in `storage.Connection` but it doesn't seem to address the headers.

WDYT? How did you determine these fixes were needed? (FWIW the `isinstance(content, six.string_types)` is needed.)

File: gcloud/storage/connection.py (hunk 2 above)
Comment (tseaver) [inline gcloud/storage/connection.py]: Again, `decode` should only be used for bytes, not text.

**Comment by dhermes** [comment]
I'm closing this out since it's being fixed upstream in `httplib2` and `oauth2client`:
https://github.com/google/oauth2client/pull/104
https://github.com/google/oauth2client/pull/136
https://github.com/jcgregorio/httplib2/pull/291
https://github.com/jcgregorio/httplib2/pull/296

**Pull request closed without merging**